devices/src/virtio/block.rs

```diff
@@ -1030,16 +1042,39 @@ mod tests {
     }
 
     #[test]
-    #[should_panic]
     fn test_invalid_event_handler() {
         let m = GuestMemory::new(&[(GuestAddress(0), 0x10000)]).unwrap();
         let (mut h, _vq) = default_test_blockepollhandler(&m);
-        // This should panic because the event is invalid.
-        h.handle_event(
+        let r = h.handle_event(
             BLOCK_EVENTS_COUNT as DeviceEventT,
             0,
             EpollHandlerPayload::Empty,
         );
+        match r {
+            Err(DeviceError::UnknownEvent { event, device }) => {
+                assert_eq!(event, BLOCK_EVENTS_COUNT as DeviceEventT);
+                assert_eq!(device, "block");
+            }
+            _ => panic!("invalid"),
+        }
+    }
+
+    #[test]
+    fn test_queue_event_read_fail_event_handler() {
+        let m = GuestMemory::new(&[(GuestAddress(0), 0x10000)]).unwrap();
+        let (mut h, _vq) = default_test_blockepollhandler(&m);
+
+        let r = h.handle_event(
```

Comment:
This is blocking (nobody is writing on the eventfd and `read` blocks indefinitely) and causing the unit tests (and therefore the coverage tests too) to fail. The test should be triggering a `read` failure so that it doesn't block. You can force one by closing the eventfd before `read`ing - it's a normal file descriptor.

```diff
diff --git a/devices/src/virtio/block.rs b/devices/src/virtio/block.rs
index 3bd3651..98ffdcb 100644
--- a/devices/src/virtio/block.rs
+++ b/devices/src/virtio/block.rs
@@ -1064,6 +1064,7 @@ mod tests {
         let m = GuestMemory::new(&[(GuestAddress(0), 0x10000)]).unwrap();
         let (mut h, _vq) = default_test_blockepollhandler(&m);
 
+        unsafe { libc::close(h.queue_evt.as_raw_fd()) };
         let r = h.handle_event(
             QUEUE_AVAIL_EVENT as DeviceEventT,
             0,
```

Reply:
Manual closing of the underlying FD of resources (by manual I mean, the resource itself thinks its FD is still open) is highly error-prone.

I've tried a similar approach and found myself in a world of problems 😅 

When closing the underlying FD, another thread of the same process can (with high probability) pick up that now-free FD number for a different newly created resource. This leads to:

1. The assumption that the FD will be unused and operations on it will fail is broken (they will instead work on another resource).
1. Operations on the original resource will unexpectedly change the state for the other resource.
1. When the original resource is released, it will actually close the FD again, thus further messing things up for the other resource.

There are a number of errors that we are not testing, especially in the more complex `devices` code, where stubbing is hard. It's fine by me if you don't test them either where doing so is not decently possible.